extension/sigv4authextension/extension.go

```diff
@@ -0,0 +1,181 @@
+// limitations under the License.
+
+package sigv4authextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/sigv4authextension"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"go.opentelemetry.io/collector/component/componenthelper"
+	"go.opentelemetry.io/collector/config/configauth"
+	"go.uber.org/zap"
+	grpcCredentials "google.golang.org/grpc/credentials"
+)
+
+// Sigv4Auth is a struct that implements the configauth.ClientAuthenticator interface. 
+// It provides the implementation for providing Sigv4 authentication for HTTP requests only.
+type Sigv4Auth struct {
+	cfg 				*Config
+	logger 				*zap.Logger
+	runtimeInfo 			string
+	componenthelper.StartFunc	// embedded default behavior to do nothing with Start()
+	componenthelper.ShutdownFunc	// embedded default behavior to do nothing with Shutdown()
+}
+
+// compile time check that the Sigv4Auth struct satisfies the configauth.ClientAuthenticator interface
+var _ configauth.ClientAuthenticator = (*Sigv4Auth)(nil)
+
+// SigningRoundTripper is a custom RoundTripper that performs AWS Sigv4.
+type SigningRoundTripper struct {
+	transport 	http.RoundTripper
+	signer 		*v4.Signer
+	region 		string
+	service 	string
+	runtimeInfo 	string
+}
+
+// RoundTripper() returns a custom SigningRoundTripper.
+func (sa *Sigv4Auth) RoundTripper(base http.RoundTripper) (http.RoundTripper, error) {
+	cfg := sa.cfg
+
+	creds, err := getCredsFromConfig(cfg)
+	if err != nil {
+		return base, err
+	}
+	if creds == nil {
+		return nil, errors.New("no AWS credentials exist")
+	}
+
+	signer := v4.NewSigner(creds)
+
+	// Create the SigningRoundTripper struct
+	rt := SigningRoundTripper {
+		transport: base,
+		signer: signer,
+		region: cfg.Region,
+		service: cfg.Service,
+		runtimeInfo: sa.runtimeInfo,
+	}
+	   
+	return &rt, nil
+}
+
+// PerRPCCredentials() is implemented to satisfy the configauth.ClientAuthenticator
+// interface but will not be implemented.
+func (sa *Sigv4Auth) PerRPCCredentials() (grpcCredentials.PerRPCCredentials, error) {
+	return nil, errors.New("Not Implemented")
+}
+
+// RoundTrip() executes a single HTTP transaction and returns an HTTP response, signing
+// the request with Sigv4.
+func (si *SigningRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqBody, err := req.GetBody()
+	if err != nil {
+		return nil, err
+	}
+    
+	content, err := ioutil.ReadAll(reqBody)
+	reqBody.Close()
+	if err != nil {
+	    	return nil, err
+	}
+	body := bytes.NewReader(content)
+    
+	// Clone request to ensure thread safety.
+	req2 := cloneRequest(req)
+    
+	// Add the runtime information to the User-Agent header of the request
```

Comment:
Call this thing "aws sdk information", `.awsSdkInfo` throughout or hard to understand what's in it

Reply:
Fixed